simpleui-server/src/queue.ts

```diff
@@ -0,0 +1,51 @@
+import { EventEmitter } from "events";
+
+
+export class Queue {
+    elements: Array<any>;
+    MAX_QUEUE_SIZE: number;
+    events: EventEmitter;
+
+    constructor() {
+        this.elements = [];
+        this.MAX_QUEUE_SIZE = 32;
+        this.events = new EventEmitter();
+    }
+
+    /**
+     * add item to back of queue
+     * @param element
+     */
+    enqueue(element: any) {
+        while (this.elements.length >= this.MAX_QUEUE_SIZE) {
+            this.dequeue();
+        }
+        this.elements.push(element);
+        this.events.emit('item_added');
+    }
+
+    /**
+     * remove item from front of queue
+     * @returns
+     */
+    dequeue() {
+        return this.isEmpty() ? "Queue Underflow" : this.elements.shift();
```

Comment:
This is slightly dangerous because the return type of this function is effectively String|ElementType (which in this case is Any, really the Queue as a whole should be templated since this queue is intended to be homogenous). What if the underlying item in the queue was a string? What if "Queue Underflow" was a valid item to put into the queue?

Reply:
I'm not sure that I follow, what do you mean by having the queue as a whole templated? Would it be templated if I generalized the class to be able to work will all types of values?